readers/multishard.cc

```diff
@@ -609,7 +609,9 @@ future<> evictable_reader_v2::fill_buffer() {
         auto* next_mf = co_await _reader->peek();
 
         // First make sure we've made progress w.r.t. _next_position_in_partition.
-        while (next_mf && _tri_cmp(_next_position_in_partition, buffer().back().position()) <= 0) {
+        // This loop becomes inifinite when next pos is a partition start.
+        // In that case progress is guranteed anyway, so skip this loop entirely.
+        while (!_next_position_in_partition.is_partition_start() && next_mf && _tri_cmp(_next_position_in_partition, buffer().back().position()) <= 0) {
```

Comment:
By the way, we access `buffer().back().position()` below right after this loop.
How do we know that the buffer isn't empty at this point?

Reply:
shouldn't that be done under `while (next_mf)`?

Reply:
> By the way, we access `buffer().back().position()` below, right after this loop. How do we know that the buffer isn't empty at that point?

See `!is_buffer_empty()` in the enclosing `if`.

Reply:
sorry, I got confused and thought we're consuming from the buffer, but we're actually filling it.